# Tablas de frecuencia

## 1. ¿Qué son y en qué consisten?

Son como el equivalente de los gráficos de barras pero en formato tabular. Esto quiere decir que nos permiten ver en una tabla la frecuencia o proporción de cada categoría dentro de la variable analizada.

Al igual que los gráficos de barras, son otra manera muy simple y muy usada de ver la distribución de una variable categórica.

## 2. ¿Cómo se obtiene una tabla de frecuencia?

Existen varias maneras de hacerlo dependiendo de la información que queramos ver en la tabla. Pero en general partimos del hecho de que nuestros datos están en un *DataFrame* de Pandas, así que las tablas de frecuencia se pueden obtener usando diferentes métodos de Pandas.

Para entender cómo usar estos diferentes métodos comencemos leyendo el set *data-prestamos.csv* usado en la lección anterior:

In [1]:
# Leer dataset
import pandas as pd

ruta = '/Users/miguel/Library/CloudStorage/GoogleDrive-miguel@codificandobits.com/My Drive/02-CODIFICANDOBITS.COM/04-Academia/01-Cursos/21-2024-03-Estadística-Descriptiva/datos/'

df = pd.read_csv(ruta + 'data-prestamos.csv')
df

,estado,nivel
0,Totalmente pagada,B
1,Cancelada (impago),C
2,Totalmente pagada,C
3,Totalmente pagada,C
4,Actual,B
...,...,...
450956,Actual,D
450957,Actual,D
450958,Actual,D
450959,Actual,D


Una manera muy sencilla de obtener una tabla de frecuencia es usando el método `describe()`:

In [2]:
df.describe()

,estado,nivel
count,450961,450961
unique,4,7
top,Actual,B
freq,321185,132370


En este caso, como las dos columnas del *DataFrame* son variables categóricas (*estado* y *nivel*), al usar `describe()` Pandas calcula parámetros estadísticos que permiten fácilmente caracterizar estas dos variables:

- *count*: por cada variable el número total de datos que tenemos
- *unique*: número de niveles por cada variable
- *top*: la categoría que más se repite en cada variable
- *freq*: la frecuencia de la categoría más común en cada variable

Si el *DataFrame* contiene variables numéricas, por defecto al usar `describe()` Pandas se enfocará en describir únicamente la variable numérica.

Por ejemplo, agreguemos una columna ficticia al *DataFrame* anterior:

In [3]:
import numpy as np

df['col_num'] = np.random.randint(0,100, (450961,))
df

,estado,nivel,col_num
0,Totalmente pagada,B,19
1,Cancelada (impago),C,39
2,Totalmente pagada,C,89
3,Totalmente pagada,C,57
4,Actual,B,48
...,...,...,...
450956,Actual,D,72
450957,Actual,D,56
450958,Actual,D,25
450959,Actual,D,54


Y ahora usemos `describe()`:

In [4]:
df.describe()

,col_num
count,450961.000000
mean,49.611102
std,28.859077
min,0.000000
25%,25.000000
50%,50.000000
75%,75.000000
max,99.000000


Así que por defecto `describe()` analizará únicamente variables numéricas. Pero podemos indicarle a Pandas que **excluya** variables numéricas usando `exclude=[np.number]`:

In [5]:
df.describe(exclude=[np.number])

,estado,nivel
count,450961,450961
unique,4,7
top,Actual,B
freq,321185,132370


Otra manera de obtener tablas de frecuencia es simplemente usando el método `value_counts()` que vimos en la lección anterior.

Por ejemplo, podemos generar una tabla con la frecuencia de cada categoría para la variable `nivel`:

In [6]:
df['nivel'].value_counts()

B    132370
C    120875
D     74277
A     72490
E     34804
F     12904
G      3241
Name: nivel, dtype: int64

Por defecto esta tabla ya estará ordenada de manera descendente. Si queremos mostrar la tabla de forma ascendente usamos `ascending=True`:

In [7]:
df['nivel'].value_counts(ascending=True)

G      3241
F     12904
E     34804
A     72490
D     74277
C    120875
B    132370
Name: nivel, dtype: int64

Y si usamos `normalize=True` podemos mostrar proporciones en lugar de conteos:

In [9]:
df['nivel'].value_counts(normalize=True, ascending=True)

G    0.007187
F    0.028614
E    0.077177
A    0.160746
D    0.164708
C    0.268039
B    0.293529
Name: nivel, dtype: float64

## 4. Ejemplo práctico

Realizar el análisis univariado de las variables *estado* y *nivel* usando tablas de frecuencia.

De nuevo, después de generar las tablas la idea es escribir una breve observación acerca del comportamiento de la variable.

Generaremos las tablas con estas características:
- Escala: proporciones
- Orden: descendente

Veamos qué ocurre con la variable *estado*:

In [10]:
df['estado'].value_counts(normalize=True)

Actual                0.712223
Totalmente pagada     0.215797
Cancelada (impago)    0.050273
Tarde                 0.021707
Name: estado, dtype: float64

En este caso las conclusiones serán similares a las obtenidas con los gráficos de barra. Sin embargo, con las tablas de frecuencia podemos tener valores más precisos.

Observaciones:

- Casi el 71% de los créditos se encuentran activos actualmente (estado = Actual)
- Un poco más del 21% de los créditos ya ha sido pagado (estado = Totalmente pagada)
- Aproximadamente un 5% de los créditos no se pueden recuperar (estado = Cancelada (impago))
- El 2.1% restante están activos pero los deudores no los han pagado a tiempo (estado = Tarde)

Veamos ahora qué sucede con la variable *nivel*:

In [11]:
tabla_nivel = df['nivel'].value_counts(normalize=True)
tabla_nivel

B    0.293529
C    0.268039
D    0.164708
A    0.160746
E    0.077177
F    0.028614
G    0.007187
Name: nivel, dtype: float64

De nuevo, las observaciones serán similares a las obtenidas en el caso de los gráficos de barras.

Imprimamos en pantalla el porcentaje total de los créditos medio-altos (categorías B, C y D) y el de los medio-bajos (categorías E, F y G):

In [12]:
print(f'Porcentaje de créditos medio-altos (B, C, D): {tabla_nivel["B":"D"].sum()}')
print(f'Porcentaje de créditos medio-bajos (E,F,G): {tabla_nivel["E":"G"].sum()}')

Porcentaje de créditos medio-altos (B, C, D): 0.7262756646361881
Porcentaje de créditos medio-bajos (E,F,G): 0.11297872765050636


Y ahora sí escribamos unas breves observaciones:
    
- Tenemos 7 diferentes niveles de crédito, siendo A los créditos con montos más altos y G con los montos más bajos
- Aproximadamente el 72% de los créditos otorgados son de niveles medio-alto (B, C, D)
- El 16% de los créditos otorgados son de niveles altos (A)
- Aproximadamente el 12% de los créditos restantes son de niveles medio-bajo (E, F, G)

## 5. Aspectos a tener en cuenta

- Las tablas de frecuencia son una forma alterna de analizar variables categóricas de manera individual
- La tabla contiene la frecuencia o proporción relativa de cada nivel dentro de la variable categórica
- A diferencia de los gráficos de barra, en las tablas de frecuencia podemos extraer valores un poco más precisos
- Y podemos usar métodos de Pandas como `describe()` o `value_counts()` para obtener estas tablas